### PR TITLE
Add result collection examples

### PR DESCRIPTION
**Pull request by SimonBoothroyd**
## Description

This PR adds a new example showing how to use the result collection objects. It also adds the missing result classes to the API docs.

## Status
- [X] Ready to go

### CODE REVIEW

File: examples/retrieving-results.ipynb [hunk 1]
```diff
@@ -0,0 +1,873 @@
...
...y='BTOTVAISQCMDLO-UHFFFAOYSA-N'), OptimizationResult(type='optimization', record_id='18435519', cmiles='[H:15][c:1]1[c:2]([c:5]([c:3]([c:6]([c:4]1[O:12][C:8]([H:18])([H:19])/[C:7](=[N:9]/[O:11][H:22])/[N:10]([H:20])[H:21])[Cl:14])[H:17])[Cl:13])[H:16]', inchi_key='BTOTVAISQCMDLO-UHFFFAOYSA-N'), OptimizationResult(type='optimization', record_id='18435520', cmiles='[H:15][c:1]1[c:2]([c:5]([c:3]([c:6]([c:4]1[O:12][C:8]([H:18])([H:19])/[C:7](=[N:9]/[O:11][H:22])/[N:10]([H:20])[H:21])[Cl:14])[H:17])[Cl:13])[H:16]', inchi_key='BTOTVAISQCMDLO-UHFFFAOYSA-N'), OptimizationResult(type='optimization', record_id='18435521', cmiles='[H:15][c:1]1[c:2]([c:5]([c:3]([c:6]([c:4]1[O:12][C:8]([H:18])([H:19])/[C:7](=[N:9]/[O:11][H:22])/[N:10]([H:20])[H:21])[Cl:14])[H:17])[Cl:13])[H:16]', inchi_key='BTOTVAISQCMDLO-UHFFFAOYSA-N'), OptimizationResult(type='optimization', record_id='18435522', cmiles='[H:15][c:1]1[c:2]([c:5]([c:3]([c:6]([c:4]1[O:12][C:8]([H:18])([H:19])/[C:7](=[N:9]/[O:11][H:22])/[N:10]([H:20])[H:21])[Cl:14])[H:17])[Cl:13])[H:16]', inchi_key='BTOTVAISQCMDLO-UHFFFAOYSA-N'), OptimizationResult(type='optimization', record_id='18435523', cmiles='[H:15][c:1]1[c:2]([c:5]([c:3]([c:6]([c:4]1[O:12]
```
Comment (jthorton) [inline examples/retrieving-results.ipynb:127]:
```suggestion
    "# Pull down the energy result records from the 'OpenFF BCC Refit Study COH v1.0' dataset.\n",
```

Comment (jthorton) [inline examples/retrieving-results.ipynb:127]: I think this should just be energy result records.

Comment (jthorton) [inline examples/retrieving-results.ipynb:128]: `energy_result_collection`

Comment (jthorton) [inline examples/retrieving-results.ipynb:147]: This just needs updating to match the torsiondrive dataset name.

Comment (jthorton) [inline examples/retrieving-results.ipynb:208]: I think this should be `n_molecules` and we should see a difference in the numbers if the molecule deduplication is working.

**Review by jthorton — APPROVED** [review]
This is fantastic, great job @SimonBoothroyd. Just a couple of typos to fix then feel free to merge!

**Comment by SimonBoothroyd** [comment]
Great thanks for catching these @jthorton - they should be fixed in 2b34047

**Pull request merged**